flytekit/core/type_engine.py

```diff
@@ -1865,7 +1873,7 @@ def _register_default_type_transformers():
             int,
             _type_models.LiteralType(simple=_type_models.SimpleType.INTEGER),
             lambda x: Literal(scalar=Scalar(primitive=Primitive(integer=x))),
-            lambda x: x.scalar.primitive.integer,
+            _check_and_convert_integer,
```

Comment:
It seems like this workflow is running with my change, which is unfortunate. Is there a compile time check though? I tried calling wf.compile() but didn't really do much. When I ran this without the change, it failed during execution with that type conversion error thats in the PR description. I ran it with pyflyte run and also just directly invoking the wf as a function.
Considering this, I'll take a look around the attribute resolve inside the promise because thats where the invariant is broken first (type annotation and underlying data mismatch) and not letting it slip out from there might be the safest if you agree. https://github.com/flyteorg/flytekit/blob/master/flytekit/core/promise.py#L86

Reply:
I see that this is for local only and remote is done in flytepropeller, the part which you linked on slack. I'm not sure then, I haven't tried this in remote yet, but I will.

Reply:
I guess the above code will fail at compile time for remote case, but we still have to ensure it also fail for local case. 
For local one, we can do conversion [here](https://github.com/flyteorg/flytekit/blob/master/flytekit/core/promise.py#L86). for remote one, we can do conversion [here](https://github.com/flyteorg/flyte/pull/4232/files?diff=unified&w=1#diff-ee7f936e440a7e043b3bc7acb4ea255ba991dea8f3144d24ab276c3a292de018R132) with the type we get from [here](https://github.com/flyteorg/flyte/pull/4232/files?diff=unified&w=1#diff-217ccf956cd0f397768572b53c52ddc5750f34085b3457b78a8ff51072a0b180R122)

Reply:
I'm working on the fix at the place you just linked. I'll have a commit ready today.

Reply:
Unfortunately this seems a little bit more complicated. If we want to avoid that failure case you posted above, we really need to know the type of the field before it was serialized. That information is not available during promise resolve time, only the type of the parameter that this value will be passed into is available. We know that it will be passed into an int, but not that it was serialized from a float. I'm working on an approach where we serialize the type information of fields of the dataclass, along with the actual dataclass value and get the type info from there when resolving the promise. Its somewhat working locally, but I have to clean it up before commiting it. 
Does that approach sound like it would be acceptable?

Reply:
I might not understand fully, but would be glad to take a look if you have the code